hca_ingest/downloader/flattener.py

```diff
@@ -152,8 +152,12 @@ def __get_concrete_process(entity: Entity) -> dict:
 
     @staticmethod
     def __get_concrete_ids(input_entities: Iterable[Entity], core_name: str, id_name: str):
+
+        def sort_by_concrete_type(entity):
+            return entity.schema.concrete_type
         concrete_ids = {}
-        for concrete_type, inputs_iter in groupby(input_entities, lambda entity: entity.schema.concrete_type):
+        for concrete_type, inputs_iter in groupby(sorted(input_entities, key=sort_by_concrete_type),
+                                                  lambda entity: entity.schema.concrete_type):
             inputs = list(inputs_iter)
             input_ids = [i.content[core_name][id_name] for i in inputs]
             input_uuids = [i.uuid for i in inputs]
```

Comment:
https://stackoverflow.com/questions/3266180/can-iterators-be-reset-in-python (Please take a look at the answer, has some insight on itertools.tee)

Since we are worried about memory usage rather, I have resorted to a simple solution: expand both variables (`input_uuids` and `input_ids`) while iterating the `inputs_iter`

Reply:
I also just tried with itertools.tee:

```python3
import time
import resource

from itertools import tee

input = {'uuid': 'dshjkfdshfljksdhfljkdshkfjsd', 'id': 'dhsjkfhdskjfhdskjfhsdjkf'}
input_1 = [input] * 5000000

# Old way
time_start = time.perf_counter()
for inputs_iter in tee(iter(input_1)):
    input_ids = [i['id'] for i in inputs_iter]
    input_uuids = [i['uuid'] for i in inputs_iter]
time_elapsed = (time.perf_counter() - time_start)
memMb=resource.getrusage(resource.RUSAGE_SELF).ru_maxrss/1024.0/1024.0
print ("%5.1f secs %5.1f MByte" % (time_elapsed,memMb))
```

With the result:
> 0.7 secs 134.2 MByte

So i will actually try using itertools.tee for the function :) and never trust slackoverflow again!